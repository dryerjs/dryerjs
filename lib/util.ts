import { plural, singular } from './js/pluralize';

export { plural, singular };

export const isNil = (value: any): value is null | undefined => value === null || value === undefined;

export const isNotNil = (value: any) => !isNil(value);

export const isObject = (value: any): value is object => typeof value === 'object';

export const isArray = (value: any) => Array.isArray(value);

export const isNotNullObject = (value: any): value is object => typeof value === 'object' && value !== null;

export const isFunction = (value: any) => typeof value === 'function';

export const isUndefined = (value: any): value is undefined => typeof value === 'undefined';

export const isTruthy = (value: any) => !!value;

export const isString = (value: any): value is string => typeof value === 'string';

export const defaultTo = <T>(value: T | null | undefined, defaultValue: T): T => {
  const shouldUseDefaultValue = isNil(value) || (typeof value === 'number' && isNaN(value));
  return shouldUseDefaultValue ? defaultValue : (value as T);
};

export const deepOmit = (object: any, omitKeys: string[]) => {
  if (typeof object !== 'object' || object === null) {
    return object;
  }

  const result = Array.isArray(object) ? [] : {};

  for (const key in object) {
    if (object.hasOwnProperty(key) && !omitKeys.includes(key)) {
      if (typeof object[key] === 'object' && object[key] !== null) {
        // Recursively omit keys from nested objects
        result[key] = deepOmit(object[key], omitKeys);
      } else {
        result[key] = object[key];
      }
    }
  }

  return result;
};

export const toPascalCase = (str: string) => str.replace(str[0], str[0].toUpperCase());

export const toCamelCase = (str: string) => str.replace(str[0], str[0].toLowerCase());

export const last = <T>(array: T[]) => array[array.length - 1];

export function memoize<T extends Function>(func: T, resolver?: (...args: any[]) => any): T {
  /* istanbul ignore if */
  if (typeof func !== 'function' || (resolver && typeof resolver !== 'function')) {
    throw new TypeError('FUNC_ERROR_TEXT');
  }

  const memoized = function (this: any, ...args: any[]) {
    const key = resolver ? resolver.apply(this, args) : String(args[0]);
    const cache = memoized.cache;
    if (cache.has(key)) return cache.get(key);
    const result = func.apply(this, args);
    cache.set(key, result);
    return result;
  };

  memoized.cache = new Map();
  return memoized as any;
}
