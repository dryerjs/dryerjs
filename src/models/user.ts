import * as graphql from 'graphql';
import { UseGuards } from '@nestjs/common';
import { IsEmail, MinLength } from 'class-validator';
import { Definition, Thunk, Filterable, ObjectId, Property, Skip, Id, Index } from 'dryerjs';
import { AdminGuard, UserGuard } from './fake-guards';

@Index({ email: 1 }, { unique: true, name: 'unique_email' })
@Index({ email: 'text', name: 'text' }, { name: 'searchable_user_index' })
@Definition({
  allowedApis: '*',
  enableTextSearch: true,
  resolverDecorators: {
    default: [UseGuards(UserGuard)],
    list: [UseGuards(AdminGuard)],
    write: [UseGuards(AdminGuard)],
    update: [UseGuards(UserGuard)],
  },
})
export class User {
  @Id()
  id: ObjectId;

  @Property()
  name: string;

  @Thunk(IsEmail(), { scopes: ['input'] })
  @Filterable(() => graphql.GraphQLString, { operators: ['eq', 'in'] })
  @Property()
  email: string;

  @Thunk(MinLength(5), { scopes: ['create'] })
  @Property({ output: Skip, update: Skip })
  password: string;
}
