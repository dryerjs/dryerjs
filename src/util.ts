export { plural, singular } from 'pluralize';

export const isNil = (value: any): value is null | undefined => value === null || value === undefined;

export const isNotNil = (value: any) => !isNil(value);

export const isObject = (value: any): value is object => typeof value === 'object';

export const isNotNullObject = (value: any): value is object => typeof value === 'object' && value !== null;

export const isFunction = (value: any) => typeof value === 'function';

export const isUndefined = (value: any): value is undefined => typeof value === 'undefined';

export const isTruthy = (value: any) => !!value;

export const isString = (value: any): value is string => typeof value === 'string';

export const defaultTo = <T>(value: T | null | undefined, defaultValue: T): T => {
    return value != null && !isNaN(value as any) ? value : defaultValue;
};

export const isEnumOfNumbers = (enumObject: any) => {
    return Object.values(enumObject).some((value: string) => '0123456789'.includes(value));
};

export const deepOmit = (object: any, omitKeys: string[]) => {
    if (typeof object !== 'object' || object === null) {
        return object;
    }

    const result = Array.isArray(object) ? [] : {};

    for (const key in object) {
        if (object.hasOwnProperty(key) && !omitKeys.includes(key)) {
            if (typeof object[key] === 'object' && object[key] !== null) {
                // Recursively omit keys from nested objects
                result[key] = deepOmit(object[key], omitKeys);
            } else {
                result[key] = object[key];
            }
        }
    }

    return result;
};

export const toPascalCase = (str: string) => str.replace(str[0], str[0].toUpperCase());

export const toCamelCase = (str: string) => str.replace(str[0], str[0].toLowerCase());

export const last = <T>(array: T[]) => array[array.length - 1];
