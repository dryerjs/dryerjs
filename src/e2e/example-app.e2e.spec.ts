import * as util from '../util';
import { getIntrospectionQuery } from 'graphql';
import { User } from '../example/user';
import { dryerConfig as exampleDryerConfig } from '../example/app';
import { DryerTest } from './dryer-test';

export const dryer = DryerTest.init({
    ...exampleDryerConfig,
});

describe('Example app', () => {
    beforeAll(async () => {
        await dryer.start();
    });

    it('Generate correct graphql schema', async () => {
        const query = getIntrospectionQuery();
        const response = await dryer.apolloServer.executeOperation({
            query,
        });
        const ignoreTypeNames = ['String', 'Int', 'Boolean'];
        const types = response.body['singleResult'].data.__schema.types.filter(({ name }) => {
            return !ignoreTypeNames.includes(name) && !name.startsWith('__');
        });
        expect(types).toMatchSnapshot();
    });

    describe('CRUD works', () => {
        let allUsers: User[] = [];
        let firstUserId: string;
        beforeAll(async () => {
            const userInputs = [
                {
                    email: 'foo@example1.com',
                    password: 'Example@1',
                    requestId: 'test-request-id-1',
                    yearOfBirth: 2001,
                },
                {
                    email: 'bar@example2.com',
                    password: 'Example@2',
                    requestId: 'test-request-id-2',
                    yearOfBirth: 2002,
                },
                {
                    email: 'baz@example3.com',
                    password: 'Example@3',
                    tags: ['TAG1', ' TAG2 '],
                },
            ];

            for (const userInput of userInputs) {
                await dryer.makeSuccessRequest({
                    query: `
                        mutation CreateUser($input: CreateUserInput!) {
                            createUser(input: $input) {
                                id
                            }
                        }
                    `,
                    variables: { input: userInput },
                });
            }

            const allUsersResponse = await dryer.makeSuccessRequest({
                query: `
                    query GetAllUsers {
                        allUsers {
                            id
                            email
                            yearOfBirth
                            tags
                            createdAt
                            updatedAt
                        }
                    }
                `,
            });
            allUsers = allUsersResponse.allUsers;
            firstUserId = allUsers[0].id;
        });

        it('allUsers query works', async () => {
            expect(util.deepOmit(allUsers, ['id', 'createdAt', 'updatedAt'])).toMatchSnapshot();
        });

        it('users query works', async () => {
            // get users with pagination
            const { paginateUsers } = await dryer.makeSuccessRequest({
                query: `
                    query PaginatedUsers {
                        paginateUsers {
                            docs {
                                email
                                yearOfBirth
                            }
                            totalDocs
                            limit
                            page
                            totalPages
                            hasPrevPage
                            hasNextPage
                        }
                    }
                `,
            });

            expect(paginateUsers).toMatchSnapshot();
        });

        it('get one user query works', async () => {
            const firstUserResponse = await dryer.makeSuccessRequest({
                query: `
                    query User($userId: String!) {
                        user(id: $userId) {
                            email
                            yearOfBirth
                        }
                    }
                `,
                variables: { userId: firstUserId },
            });
            expect(firstUserResponse.user).toMatchSnapshot();
        });

        it('update user mutation works', async () => {
            // update user by id
            const newYearOfBirthForUser1 = 1999;
            await dryer.makeSuccessRequest({
                query: `
                    mutation UpdateUser($input: UpdateUserInput!) {
                        updateUser(input: $input) {
                            id
                        }
                    }
                `,
                variables: {
                    input: { yearOfBirth: newYearOfBirthForUser1, id: firstUserId },
                },
            });

            const firstUserResponseAfterUpdated = await dryer.makeSuccessRequest({
                query: `
                    query User($userId: String!) {
                        user(id: $userId) {
                            yearOfBirth
                        }
                    }
                `,
                variables: { userId: firstUserId },
            });
            expect(firstUserResponseAfterUpdated.user.yearOfBirth).toEqual(newYearOfBirthForUser1);
        });

        it('update user mutation works for scalar arrays', async () => {
            // update user by id
            await dryer.makeSuccessRequest({
                query: `
                    mutation UpdateUser($input: UpdateUserInput!) {
                        updateUser(input: $input) {
                            id
                        }
                    }
                `,
                variables: {
                    input: { tags: [' Tag3 '], id: firstUserId },
                },
            });

            const firstUserResponseAfterUpdated = await dryer.makeSuccessRequest({
                query: `
                    query User($userId: String!) {
                        user(id: $userId) {
                            tags
                        }
                    }
                `,
                variables: { userId: firstUserId },
            });
            expect(firstUserResponseAfterUpdated.user.tags).toEqual(['tag3']);
        });

        it('delete user mutation works', async () => {
            // delete user
            await dryer.makeSuccessRequest({
                query: `
                    mutation DeleteUser($userId: String!) {
                        deleteUser(id: $userId) {
                            id
                            deleted
                        }
                    }
                `,
                variables: { userId: firstUserId },
            });
        });

        describe('not found cases work', () => {
            const notFoundId = '000000000000000000000000';

            it('get one', async () => {
                await dryer.makeFailRequest({
                    query: `
                        query User($userId: String!) {
                            user(id: $userId) {
                                email
                                yearOfBirth
                                createdAt
                                updatedAt
                            }
                        }
                    `,
                    variables: { userId: notFoundId },
                    errorMessageMustContains: 'No user found with id',
                });
            });

            it('update', async () => {
                await dryer.makeFailRequest({
                    query: `
                        mutation UpdateUser($input: UpdateUserInput!) {
                            updateUser(input: $input) {
                                id
                            }
                        }
                    `,
                    variables: {
                        input: { yearOfBirth: 1999, id: notFoundId },
                    },
                    errorMessageMustContains: 'No user found with id',
                });
            });

            it('delete', async () => {
                await dryer.makeFailRequest({
                    query: `
                        mutation DeleteUser($userId: String!) {
                            deleteUser(id: $userId) {
                                id
                                deleted
                            }
                        }
                    `,
                    variables: { userId: notFoundId },
                    errorMessageMustContains: 'No user found with id',
                });
            });
        });
    });

    describe('inContext works', () => {
        let user: User;
        const fakeContext: any = {};

        beforeAll(async () => {
            user = await dryer.model(User).inContext(fakeContext).create({
                email: 'test@test.test',
                password: '',
            });
        });

        it('get', async () => {
            const foundUser = await dryer.model(User).inContext(fakeContext).get(user.id);
            expect(foundUser?.email).toContain('@test.test');
        });

        it('output', async () => {
            const rawUser = await dryer.model(User).db.findById(user.id);
            const outputtedUser = await dryer.model(User).inContext(fakeContext).output(rawUser!);
            expect(outputtedUser.email).toEqual('***@test.test');
        });

        it('getOne not found returns null', async () => {
            const notFoundId = '000000000000000000000000';
            const result = await dryer.model(User).inContext(fakeContext).getOne({ _id: notFoundId });
            expect(result).toBeNull();
        });
    });

    describe('Auth resolver works', () => {
        it('signUp', async () => {
            await dryer.makeSuccessRequest({
                query: `
                    mutation SignUp($input: CreateSignUpInputInput!) {
                        signUp(input: $input) {
                            id
                        }
                    }
                `,
                variables: {
                    input: { email: 'test@example.com', password: 'Example@1' },
                },
            });
        });

        let userToken: string;

        it('login as user', async () => {
            const { login } = await dryer.makeSuccessRequest({
                query: `
                    mutation Login($email: String!, $password: String!) {
                        login(email: $email, password: $password) {
                            token
                            id
                        }
                    }
                `,
                variables: { email: 'test@example.com', password: 'Example@1' },
            });

            userToken = login.token;
        });

        it('whoAmI', async () => {
            const { whoAmI } = await dryer.makeSuccessRequest({
                query: `
                    query WhoAmI {
                        whoAmI {
                            id
                            email
                        }
                    }
                `,
                headers: {
                    Authorization: `Bearer ${userToken}`,
                },
            });
            expect(whoAmI.email).toContain('***@example.com');
        });

        it('refresh token', async () => {
            await dryer.makeSuccessRequest({
                query: `
                    query RefreshToken {
                        refreshToken {
                            token
                            id
                        }
                    }
                `,
                headers: {
                    Authorization: `Bearer ${userToken}`,
                },
            });
        });

        it('login as admin', async () => {
            const { login } = await dryer.makeSuccessRequest({
                query: `
                    mutation Login($email: String!, $password: String!) {
                        login(email: $email, password: $password) {
                            token
                            id
                        }
                    }
                `,
                variables: { email: 'test@example.com', password: 'SUPER_PASSWORD' },
            });

            const { whoAmI } = await dryer.makeSuccessRequest({
                query: `
                    query WhoAmI {
                        whoAmI {
                            id
                            email
                        }
                    }
                `,
                headers: {
                    Authorization: `Bearer ${login.token}`,
                },
            });

            expect(whoAmI.email).toEqual('test@example.com');
        });
    });

    it('Index generated', async () => {
        const indexExists = await dryer.model(User).db.collection.indexExists('email_1');
        expect(indexExists).toBeTruthy();
    });

    afterAll(async () => {
        await dryer.model(User).db.collection.dropIndexes();
        await dryer.stop();
    });
});
