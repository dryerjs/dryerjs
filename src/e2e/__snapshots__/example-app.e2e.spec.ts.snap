// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example app CRUD works allUsers query works 1`] = `
[
  {
    "email": "***@example1.com",
    "yearOfBirth": 2001,
  },
  {
    "email": "***@example2.com",
    "yearOfBirth": 2002,
  },
  {
    "email": "***@example3.com",
    "yearOfBirth": 1990,
  },
]
`;

exports[`Example app CRUD works get one user query works 1`] = `
{
  "email": "***@example1.com",
  "yearOfBirth": 2001,
}
`;

exports[`Example app CRUD works users query works 1`] = `
{
  "docs": [
    {
      "email": "***@example1.com",
      "yearOfBirth": 2001,
    },
    {
      "email": "***@example2.com",
      "yearOfBirth": 2002,
    },
    {
      "email": "***@example3.com",
      "yearOfBirth": 1990,
    },
  ],
  "hasNextPage": false,
  "hasPrevPage": false,
  "limit": 10,
  "page": 1,
  "totalDocs": 3,
  "totalPages": 1,
}
`;

exports[`Example app Generate correct graphql schema 1`] = `
[
  {
    "description": null,
    "enumValues": null,
    "fields": [
      {
        "args": [
          {
            "defaultValue": null,
            "description": null,
            "name": "id",
            "type": {
              "kind": "NON_NULL",
              "name": null,
              "ofType": {
                "kind": "SCALAR",
                "name": "String",
                "ofType": null,
              },
            },
          },
        ],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "user",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "OBJECT",
            "name": "User",
            "ofType": null,
          },
        },
      },
      {
        "args": [
          {
            "defaultValue": null,
            "description": null,
            "name": "skip",
            "type": {
              "kind": "SCALAR",
              "name": "Int",
              "ofType": null,
            },
          },
          {
            "defaultValue": null,
            "description": null,
            "name": "take",
            "type": {
              "kind": "SCALAR",
              "name": "Int",
              "ofType": null,
            },
          },
        ],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "paginateUsers",
        "type": {
          "kind": "OBJECT",
          "name": "PaginatedUsers",
          "ofType": null,
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "allUsers",
        "type": {
          "kind": "LIST",
          "name": null,
          "ofType": {
            "kind": "NON_NULL",
            "name": null,
            "ofType": {
              "kind": "OBJECT",
              "name": "User",
              "ofType": null,
            },
          },
        },
      },
    ],
    "inputFields": null,
    "interfaces": [],
    "kind": "OBJECT",
    "name": "Query",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": [
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "id",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "email",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "yearOfBirth",
        "type": {
          "kind": "SCALAR",
          "name": "Int",
          "ofType": null,
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "lastLoggedInAt",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "createdAt",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "phone",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "OBJECT",
            "name": "Phone",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "addresses",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "LIST",
            "name": null,
            "ofType": {
              "kind": "OBJECT",
              "name": "Address",
              "ofType": null,
            },
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "status",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "ENUM",
            "name": "UserStatus",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "updatedAt",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
    ],
    "inputFields": null,
    "interfaces": [],
    "kind": "OBJECT",
    "name": "User",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": [
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "phoneNumber",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "countryPrefix",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
    ],
    "inputFields": null,
    "interfaces": [],
    "kind": "OBJECT",
    "name": "Phone",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": [
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "id",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "streetAddress",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "postalCode",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
    ],
    "inputFields": null,
    "interfaces": [],
    "kind": "OBJECT",
    "name": "Address",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": [
      {
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "ACTIVE",
      },
      {
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "INACTIVE",
      },
    ],
    "fields": null,
    "inputFields": null,
    "interfaces": null,
    "kind": "ENUM",
    "name": "UserStatus",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": [
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "docs",
        "type": {
          "kind": "LIST",
          "name": null,
          "ofType": {
            "kind": "NON_NULL",
            "name": null,
            "ofType": {
              "kind": "OBJECT",
              "name": "User",
              "ofType": null,
            },
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "totalDocs",
        "type": {
          "kind": "SCALAR",
          "name": "Int",
          "ofType": null,
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "page",
        "type": {
          "kind": "SCALAR",
          "name": "Int",
          "ofType": null,
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "limit",
        "type": {
          "kind": "SCALAR",
          "name": "Int",
          "ofType": null,
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "hasPrevPage",
        "type": {
          "kind": "SCALAR",
          "name": "Boolean",
          "ofType": null,
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "hasNextPage",
        "type": {
          "kind": "SCALAR",
          "name": "Boolean",
          "ofType": null,
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "totalPages",
        "type": {
          "kind": "SCALAR",
          "name": "Int",
          "ofType": null,
        },
      },
    ],
    "inputFields": null,
    "interfaces": [],
    "kind": "OBJECT",
    "name": "PaginatedUsers",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": [
      {
        "args": [
          {
            "defaultValue": null,
            "description": null,
            "name": "input",
            "type": {
              "kind": "NON_NULL",
              "name": null,
              "ofType": {
                "kind": "INPUT_OBJECT",
                "name": "CreateUserInput",
                "ofType": null,
              },
            },
          },
        ],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "createUser",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "OBJECT",
            "name": "User",
            "ofType": null,
          },
        },
      },
      {
        "args": [
          {
            "defaultValue": null,
            "description": null,
            "name": "id",
            "type": {
              "kind": "NON_NULL",
              "name": null,
              "ofType": {
                "kind": "SCALAR",
                "name": "String",
                "ofType": null,
              },
            },
          },
          {
            "defaultValue": null,
            "description": null,
            "name": "input",
            "type": {
              "kind": "NON_NULL",
              "name": null,
              "ofType": {
                "kind": "INPUT_OBJECT",
                "name": "UpdateUserInput",
                "ofType": null,
              },
            },
          },
        ],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "updateUser",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "OBJECT",
            "name": "User",
            "ofType": null,
          },
        },
      },
      {
        "args": [
          {
            "defaultValue": null,
            "description": null,
            "name": "id",
            "type": {
              "kind": "NON_NULL",
              "name": null,
              "ofType": {
                "kind": "SCALAR",
                "name": "String",
                "ofType": null,
              },
            },
          },
        ],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "deleteUser",
        "type": {
          "kind": "OBJECT",
          "name": "DeleteResponse",
          "ofType": null,
        },
      },
    ],
    "inputFields": null,
    "interfaces": [],
    "kind": "OBJECT",
    "name": "Mutation",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": null,
    "inputFields": [
      {
        "defaultValue": null,
        "description": null,
        "name": "email",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "password",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "yearOfBirth",
        "type": {
          "kind": "SCALAR",
          "name": "Int",
          "ofType": null,
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "requestId",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "phone",
        "type": {
          "kind": "INPUT_OBJECT",
          "name": "CreatePhoneInput",
          "ofType": null,
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "addresses",
        "type": {
          "kind": "LIST",
          "name": null,
          "ofType": {
            "kind": "INPUT_OBJECT",
            "name": "CreateAddressInput",
            "ofType": null,
          },
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "status",
        "type": {
          "kind": "ENUM",
          "name": "UserStatus",
          "ofType": null,
        },
      },
    ],
    "interfaces": null,
    "kind": "INPUT_OBJECT",
    "name": "CreateUserInput",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": null,
    "inputFields": [
      {
        "defaultValue": null,
        "description": null,
        "name": "phoneNumber",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "countryPrefix",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
    ],
    "interfaces": null,
    "kind": "INPUT_OBJECT",
    "name": "CreatePhoneInput",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": null,
    "inputFields": [
      {
        "defaultValue": null,
        "description": null,
        "name": "streetAddress",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "postalCode",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
    ],
    "interfaces": null,
    "kind": "INPUT_OBJECT",
    "name": "CreateAddressInput",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": null,
    "inputFields": [
      {
        "defaultValue": null,
        "description": null,
        "name": "yearOfBirth",
        "type": {
          "kind": "SCALAR",
          "name": "Int",
          "ofType": null,
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "requestId",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "phone",
        "type": {
          "kind": "INPUT_OBJECT",
          "name": "UpdatePhoneInput",
          "ofType": null,
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "addresses",
        "type": {
          "kind": "LIST",
          "name": null,
          "ofType": {
            "kind": "INPUT_OBJECT",
            "name": "UpdateAddressInput",
            "ofType": null,
          },
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "status",
        "type": {
          "kind": "ENUM",
          "name": "UserStatus",
          "ofType": null,
        },
      },
    ],
    "interfaces": null,
    "kind": "INPUT_OBJECT",
    "name": "UpdateUserInput",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": null,
    "inputFields": [
      {
        "defaultValue": null,
        "description": null,
        "name": "phoneNumber",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "countryPrefix",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
    ],
    "interfaces": null,
    "kind": "INPUT_OBJECT",
    "name": "UpdatePhoneInput",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": null,
    "inputFields": [
      {
        "defaultValue": null,
        "description": null,
        "name": "streetAddress",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
      {
        "defaultValue": null,
        "description": null,
        "name": "postalCode",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
    ],
    "interfaces": null,
    "kind": "INPUT_OBJECT",
    "name": "UpdateAddressInput",
    "possibleTypes": null,
  },
  {
    "description": null,
    "enumValues": null,
    "fields": [
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "deleted",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "Boolean",
            "ofType": null,
          },
        },
      },
      {
        "args": [],
        "deprecationReason": null,
        "description": null,
        "isDeprecated": false,
        "name": "id",
        "type": {
          "kind": "NON_NULL",
          "name": null,
          "ofType": {
            "kind": "SCALAR",
            "name": "String",
            "ofType": null,
          },
        },
      },
    ],
    "inputFields": null,
    "interfaces": [],
    "kind": "OBJECT",
    "name": "DeleteResponse",
    "possibleTypes": null,
  },
]
`;
